codex/marketplace.md

```diff
@@ -0,0 +1,209 @@
+The marketplace is a a smart contract on a EVM compatiable blockchain.
+The smart contract SHOULD handle storage request, storage slot state, 
+storage provider rewards, storage provider collateral, and storage proof state. 
+
+To create a request to store a dataset on the Codex network,
+client nodes MUST split the dataset into data chunks, $(c_1, c_2, c_3, \ldots, c_{20})$.
+Using a erasure coding technique, 
+the data chunks are encoded and placed into separate slots.
+The RECCOMENDED erasure coding technique is a [Reed-Soloman algorithm](https://hackmd.io/FB58eZQoTNm-dnhu0Y1XnA).
+
+The requester, client node, SHOULD submit a transaction with the desired request parameters.
+The requester MUST provide a `duration` value for the storage request along with the appropriate `reward`,
+payment for request. 
+Once a request is created via the transaction, 
+all slots MUST be filled by storage providers before the request is officially started.
+If the request does not attract enough storage providers after a pre-defined network timeout,
+which is defined by the Codex node as `expiry`,
+the request MUST be canceled.
+If canceled, `collateral` SHOULD be returned to any storage providers and 
+`reward` returned to the requester.
+The requester MAY create a new request with different values to restart the process.
+
+The storage request SHOULD initiate a new storage contract with the following parameters:
+
+```solidity
+
+function requestStorage() public {
+  // the Codex node requesting storage
+  address client;
+
+  // content identifier
+  string cid;
+
+  // merkle root of the dataset, used to verify storage proofs
+  byte32 merkleRoot;
+
+  // tokens from the requester to reward storage providers
+  uint256 reward;
+
+  // amount of tokens required for collateral by storage providers
+  UInt256 collateral;
+
+  // frequency that proofs are checked by validator nodes
+  uint256 proofProbability;
+
+  // amount of desired time for stoage request
+  uint256 duration;
+
+  // the number of requested slots
+  uint64 slots;
+
+  // amount of storage per slot
+  uint256 slotSize;
+
+  // Amount of time before request expires
+  uint256 expiry;
+
+  // random value to differentiate from other requests
+  byte32 nonce;
+
+}
+
+```
+
+`cid` 
+
+An identifier used to locate the dataset
+- MUST be a sha2-256 hash (length of 32 bytes),
+- MUST be generated by the requester
+
+`reward`
+
+- SHOULD be a token known to the network.
+- it MUST be paid directly to storage providers who fill slots for the `duration` of the request
+
+`collateral`
+
+All storage providers MUST provide token collateral before being able to fill a storage slot.
+The collateral is used to punish a storage provider for being an malicious or unreliable.
+The following is related to storage provider who has offered `collateral`
+
+If a storage provider, filling a slot,
+fails to provide enough proofs of storage, the `collateral` MUST be forfeited.
```

Comment:
Again there is the question of "what are we specing here": 

> the `collateral` MUST be forfeited

It is not like the nodes MUST "forfeit" their collateral by doing some smart contract call. This is handled by the smart contract in the form of slashing. As smart contracts can not "act automatically" (eq. they automatically slash collateral when nodes miss their proof), these actions need to be triggered by somebody submitting transaction (and paying for it), we could potentially talk about who SHOULD/MUST do this triggering. But honestly I would leave that to separate section where we should spec how Validator node should behave.

```suggestion
miss the critical amount of submitted proofs of the storage, the `collateral` will be slashed. If a critical amount of `collateral` is slashed, then the slot is vacated and the repair process kicks in.
```

Reply:
I agree forfeited may be the wrong term to use here. From my understanding slashing refers to a user losing a portion of their tokens. This should be mention as well as total lost of collateral when maxNumOfSlashing occurs.

Reply:
I think `forfeited` is an appropriate term here, I also think that leaving the mechanics of how this is accomplished open ended is the right approach here. The fact that a contract cannot trigger actions internally and has to be externally initiated is an implementation detail, and a limitation of the underlying execution environment and should be captured in a different spec, possibly by the validator spec.

IMO, this spec should be generic enough, that it can be implemented on top of a different execution environment altogether.

So, I overall agree with the current wording.